File: backend/src/main/java/ma/investpro/repository/ConventionRepository.java
package ma.investpro.repository;

import ma.investpro.entity.Convention;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Repository
public interface ConventionRepository extends JpaRepository<Convention, Long> {

    Optional<Convention> findByCode(String code);

    List<Convention> findByActifTrue();

    @Query("SELECT c FROM Convention c WHERE c.actif = true " +
           "AND c.dateDebut <= :date " +
           "AND (c.dateFin IS NULL OR c.dateFin >= :date)")
    List<Convention> findActiveAtDate(@Param("date") LocalDate date);

    boolean existsByCode(String code);

    @Query("SELECT c FROM Convention c WHERE " +
           "LOWER(c.code) LIKE LOWER(CONCAT('%', :search, '%')) OR " +
           "LOWER(c.libelle) LIKE LOWER(CONCAT('%', :search, '%'))")
    List<Convention> search(@Param("search") String search);
}

File: backend/src/main/kotlin/ma/investpro/repository/ConventionRepository.kt
package ma.investpro.repository

import ma.investpro.entity.Convention
import ma.investpro.entity.StatutConvention
import ma.investpro.entity.TypeConvention
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.stereotype.Repository
import java.time.LocalDate

@Repository
interface ConventionRepository : JpaRepository<Convention, Long> {

    // Recherche par code unique
    fun findByCode(code: String): Convention?

    // Recherche par numéro unique
    fun findByNumero(numero: String): Convention?

    // Recherche par type
    fun findByTypeConvention(type: TypeConvention): List<Convention>

    // Recherche par statut
    fun findByStatut(statut: StatutConvention): List<Convention>

    // Recherche par statut multiple
    fun findByStatutIn(statuts: List<StatutConvention>): List<Convention>

    // Recherche des conventions actives (pas annulées)
    fun findByActifTrue(): List<Convention>

    // Recherche des sous-conventions d'une convention parente
    fun findByParentConventionId(parentId: Long): List<Convention>

    // Recherche des conventions racine (sans parent)
    fun findByParentConventionIsNull(): List<Convention>

    // Recherche par période
    @Query("SELECT c FROM Convention c WHERE c.dateDebut <= :dateFin AND (c.dateFin IS NULL OR c.dateFin >= :dateDebut)")
    fun findByPeriode(dateDebut: LocalDate, dateFin: LocalDate): List<Convention>

    // Recherche des conventions en cours
    @Query("SELECT c FROM Convention c WHERE c.statut = 'EN_COURS' AND c.dateDebut <= CURRENT_DATE AND (c.dateFin IS NULL OR c.dateFin >= CURRENT_DATE)")
    fun findConventionsEnCours(): List<Convention>

    // Recherche des conventions expirées ou en retard
    @Query("SELECT c FROM Convention c WHERE c.statut IN ('EN_COURS', 'VALIDEE') AND c.dateFin < CURRENT_DATE")
    fun findConventionsExpirees(): List<Convention>

    // Recherche des conventions verrouillées
    fun findByIsLockedTrue(): List<Convention>

    // Recherche par libellé (like)
    fun findByLibelleContainingIgnoreCase(libelle: String): List<Convention>

    // Vérifier si un code existe déjà
    fun existsByCode(code: String): Boolean

    // Vérifier si un numéro existe déjà
    fun existsByNumero(numero: String): Boolean

    // Compter par statut
    fun countByStatut(statut: StatutConvention): Long

    // Compter par type
    fun countByTypeConvention(type: TypeConvention): Long
}
